### PR TITLE
Sample properties

### LINKED ISSUE #82: Add sample properties windows
**This issue is a feature request**

A sample can either be a triangle mesh or a volume. The properties of the sample should be discoverable from within the software.

1. Create a "Sample Properties" windows to show the information related to the current sample. The following info should be displayed the window:
      - Memory used for data only not classes (for volume and mesh)
      - Dimension (volume only)
      - Voxel size (volume only)
      - Min and Max value (volume only)
      - Faces (mesh only)
      - Vertices (mesh only)

2. This window should be docked at the bottom (DockFlag.Bottom) in the dock manager. 
3. The sample info should update appropriately if the sample is changed while the window is opened. The window should also properly handle the case when the sample is removed (using the undo button) while the window is opened 
4. This window should be accessible via the menu **View > Other Windows > Sample Properites** with shortcut **Ctrl+Shift+I** 
5. The unit test and documentation should be updated appropriately

### PR DESCRIPTION
**Pull request by RabiyaF**
Closes #82 

Added the functionality to view the sample properties of the imported model. 

- [x] Added a manager to display the window with information for mesh and volume samples.
- [x] Added the menu bar access of widget through **View > Other Windows > Sample Properties**
- [x] Added the keyboard shortcut of **Ctrl+Shift+I** for quick access to the sample information.
- [x] Memory (in Mb), (# of) Faces and (# of) Vertices are displayed for mesh samples.
- [x] Memory (in Mb), Dimension, Voxel size, Minimum intensity and Maximum intensity are displayed for volume samples.
- [x] The sample properties will update as the sample is changed/removed.

* **Other information:**
* The Sample properties window is docked at the bottom in the dock manager.
* In case no sample is imported, the properties widget will display a Memory of 0.
* Unit tests have been added to test_widget.py to test the new feature for (2) meshes, (2) volumes and (1) no sample scenario. 
* Test have been updated in test_ui.py to test the app UI when a mesh sample is uploaded. 
* The documentation has been updated to include the description of the new feature.

### CODE REVIEW

File: docs/add_sample.rst [hunk 1]
```diff
@@ -44,6 +44,28 @@ the current one or be combined with it.
...
+For mesh samples- the memory (in mb), faces and vertices are displayed. 
+
+.. image:: images/mesh_sample_properties.png
+   :scale: 80
+   :alt: Mesh sample properties dialog 
+   :align: center
+
+For volume samples- the memory (in mb), dimensions, voxel size, minimum intensity and maximum intensity are displayed. 
+
+.. image:: images/volume_sample_properties.png
```
Comment (StephenNneji) [inline docs/add_sample.rst:61]: Change the extension of the 2 images from .PNG to .png. On Linux, the two are not he same

File: sscanss/app/dialogs/managers.py [hunk 2]
```diff
@@ -943,3 +947,153 @@ def moveDetectorsButtonClicked(self):
...
+    def addSamplePropertyValues(self):
+        """Adds the sample property values to the table"""
+        if isinstance(self.parent_model.sample, Mesh):
+            self._addMeshPropertyValues()
+        elif isinstance(self.parent_model.sample, Volume):
+            self._addVolumePropertyValues()
+        elif self.parent_model.sample is None:
+            self._add_property_value('0', row=0)
+
+    def _addItem(self, item, is_label, row):
```
Comment (StephenNneji) [inline sscanss/app/dialogs/managers.py:1079]: Would it not ne simpler to do 

`def _addItem(self, row, label, value)`

What is the benefit of having code to set the label then code to set the value?

File: sscanss/app/dialogs/managers.py [hunk 3]
```diff
@@ -943,3 +947,153 @@ def moveDetectorsButtonClicked(self):
...
+        col = 0 if is_label else 1
+        item = str(item)
+        entry_item = QtWidgets.QTableWidgetItem(item)
+        entry_item.setTextAlignment(QtCore.Qt.AlignCenter)
+        self.sample_property_table.setItem(row, col, entry_item)
+
+    _add_property_label = partialmethod(_addItem, is_label=True)
+    _add_property_value = partialmethod(_addItem, is_label=False)
+
+    def closeEvent(self, event):
```
Comment (StephenNneji) [inline sscanss/app/dialogs/managers.py:1098]: This is not needed. You only need to override the close event handler if you are changing its behaviour like performing some clean-up

File: tests/test_ui.py [hunk 4]
```diff
@@ -179,6 +196,24 @@ def addSample(self):
         widget_2 = self.getDockedWidget(self.window.docks, InsertPrimitiveDialog.dock_flag)
         self.assertIsNot(widget, widget_2)
         old_vertex_count = len(self.model.sample.vertices)
+
+        self.window.sample_properties_dialog_action.trigger()
```
Comment (StephenNneji) [inline tests/test_ui.py:200]: The dialog is already open. You don need to trigger the action again or assert the instance

File: tests/test_ui.py [hunk 5]
```diff
@@ -179,6 +196,24 @@ def addSample(self):
...
+        sample_properties_widget = self.getDockedWidget(self.window.docks, SampleProperties.dock_flag)
+        self.assertTrue(isinstance(sample_properties_widget, SampleProperties))
+        bytes_used = self.model.sample.vertices.nbytes
+        bytes_to_mb_factor = 1 / (1024**2)
```
Comment (StephenNneji) [inline tests/test_ui.py:204]: you already defined this above

File: tests/test_ui.py [hunk 6]
```diff
@@ -179,6 +196,24 @@ def addSample(self):
...
+        memory = bytes_used * bytes_to_mb_factor
+        memory_formatted = f'{memory:.4f}'
```
Comment (StephenNneji) [inline tests/test_ui.py:206]: you don't need this variable. Use the f-string directly its already clear that you are formatting the memory variable

File: tests/test_ui.py [hunk 7]
```diff
@@ -179,6 +196,24 @@ def addSample(self):
...
+        self.assertEqual('Memory (MB)', sample_properties_widget.sample_property_table.item(0, 0).text())
+        self.assertEqual(memory_formatted, sample_properties_widget.sample_property_table.item(0, 1).text())
+        num_indices, = self.model.sample.indices.shape
```
Comment (StephenNneji) [inline tests/test_ui.py:209]: also unnecessary variable you can use `self.model.sample.indices.shape[0]` in short statements

File: tests/test_ui.py [hunk 8]
```diff
@@ -179,6 +196,24 @@ def addSample(self):
...
+        num_faces = num_indices // 3
+        self.assertEqual('Faces', sample_properties_widget.sample_property_table.item(1, 0).text())
+        self.assertEqual(str(num_faces), sample_properties_widget.sample_property_table.item(1, 1).text())
+        num_vertices, _ = self.model.sample.vertices.shape
```
Comment (StephenNneji) [inline tests/test_ui.py:213]: also unnecessary variable you can use `self.model.sample.indices.shape[0]` or if you keep the `num_vertices` above then there is no need to redefine it here

File: tests/test_ui.py [hunk 9]
```diff
@@ -170,6 +170,23 @@ def addSample(self):
         QTest.mouseClick(widget.create_primitive_button, Qt.LeftButton)
         self.assertIsNotNone(self.model.sample)
 
+        self.window.sample_properties_dialog_action.trigger()
+        sample_properties_widget = self.getDockedWidget(self.window.docks, SampleProperties.dock_flag)
+        self.assertTrue(isinstance(sample_properties_widget, SampleProperties))
+        bytes_used = self.model.sample.vertices.nbytes
```
Comment (StephenNneji) [inline tests/test_ui.py:176]: unnecessary variable `self.model.sample.vertices.nbytes` is clear

File: tests/test_widgets.py [hunk 10]
```diff
@@ -2072,5 +2073,111 @@ def testValidation(self):
...
+        memory_formatted = f'{memory:.4f}'
+        self.assertEqual('Memory (MB)', self.dialog.sample_property_table.item(0, 0).text())
+        self.assertEqual(memory_formatted, self.dialog.sample_property_table.item(0, 1).text())
+
+        num_indices, = self.mesh.indices.shape
+        num_faces = num_indices // 3
+        self.assertEqual('Faces', self.dialog.sample_property_table.item(1, 0).text())
+        self.assertEqual(str(num_faces), self.dialog.sample_property_table.item(1, 1).text())
+
+        num_vertices, _ = self.mesh.vertices.shape
```
Comment (StephenNneji) [inline tests/test_widgets.py:2122]: `num_vertices` is the same as above no need to redefine

File: sscanss/app/dialogs/managers.py [hunk 11]
```diff
@@ -943,3 +947,153 @@ def moveDetectorsButtonClicked(self):
...
+
+        :param label_dict: the dictionary containing the property labels
+        :type label_dict: dict
+        """
+        for row, label_tag in label_dict.items():
+            self._add_property_label(label_tag, row=row)
+
+    def _addMeshPropertyValues(self):
+        """Adds the mesh property values to the table"""
+        memory_str = self._getMemoryUsed()
```
Comment (StephenNneji) [inline sscanss/app/dialogs/managers.py:1013]: Unnecessary variable. _getMemoryUsed() is already very clear

File: sscanss/app/dialogs/managers.py [hunk 12]
```diff
@@ -943,3 +947,153 @@ def moveDetectorsButtonClicked(self):
...
+        self._add_property_value(memory_str, row=0)
+
+        num_indices, = self.parent_model.sample.indices.shape
+        num_faces = num_indices // 3
+        self._add_property_value(num_faces, row=1)
+
+        num_vertices, _ = self.parent_model.sample.vertices.shape
```
Comment (StephenNneji) [inline sscanss/app/dialogs/managers.py:1020]: Already defined above

**Review by StephenNneji — COMMENTED** [review]
Good job. Please address comments

File: docs/add_sample.rst (hunk 1 above)
Comment (RabiyaF) [reply docs/add_sample.rst:61]: converted the image extension from PNG to png.

File: sscanss/app/dialogs/managers.py (hunk 11 above)
Comment (RabiyaF) [reply sscanss/app/dialogs/managers.py:1013]: Unnecessary variable removed.

File: sscanss/app/dialogs/managers.py (hunk 3 above)
Comment (RabiyaF) [reply sscanss/app/dialogs/managers.py:1098]: Unnecessary method removed.

File: sscanss/app/dialogs/managers.py (hunk 2 above)
Comment (RabiyaF) [reply sscanss/app/dialogs/managers.py:1079]: Refactored code to remove the extra methods

File: tests/test_ui.py (hunk 9 above)
Comment (RabiyaF) [reply tests/test_ui.py:176]: bytes_used removed from test_widget.py and test_ui.py

File: tests/test_ui.py (hunk 5 above)
Comment (RabiyaF) [reply tests/test_ui.py:204]: second definition of bytes_to_mb_factor removed from test_ui.py

File: tests/test_ui.py (hunk 6 above)
Comment (RabiyaF) [reply tests/test_ui.py:206]: memory_formatted variable removed from test_widget.py and test_ui.py

File: tests/test_ui.py (hunk 4 above)
Comment (RabiyaF) [reply tests/test_ui.py:200]: Removed the redundant assert statement and the unnecessary widget trigger.

File: tests/test_ui.py (hunk 7 above)
Comment (RabiyaF) [reply tests/test_ui.py:209]: Removed num_indices in variable clean up.

File: sscanss/app/dialogs/managers.py [hunk 13]
```diff
@@ -943,3 +947,153 @@ def moveDetectorsButtonClicked(self):
...
+        self.main_layout = QtWidgets.QVBoxLayout()
+        self.sample_property_table = QtWidgets.QTableWidget()
+        self.sample_property_table.horizontalHeader().setSectionResizeMode(QtWidgets.QHeaderView.Stretch)
+        self.sample_property_table.verticalHeader().hide()
+        self.sample_property_table.setColumnCount(2)
+        self.sample_property_table.setRowCount(5)
+        self.sample_property_table.setAlternatingRowColors(True)
+        self.sample_property_table.setMinimumHeight(165)
+        self.sample_property_table.setMaximumHeight(210)
+        stylesheet = "::section{Background-color:rgb(100,100,100); font-weight: bold; color: rgb(255, 255, 255);}"
```
[the remark this answers is not in the document]
Comment (RabiyaF) [reply sscanss/app/dialogs/managers.py:974]: Custom table header moved to style sheets.

**Comment by RabiyaF** [comment]
The image extensions were not pushed correctly. I am working to fix it.

**Comment by RabiyaF** [comment]
https://github.com/desktop/desktop/issues/11087

**Review by StephenNneji — COMMENTED** [review]
Good job 👍

**Pull request merged**